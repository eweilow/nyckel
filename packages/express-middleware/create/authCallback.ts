import { RequestHandler } from "express";
import { NyckelExpressMiddlewareOptions } from "../middleware";
import { asyncHandler } from "..";
import {
  concatUrl,
  requestToken,
  verifyCSRFPair
} from "@nyckel/authentication";

export function createCallbackHandler(
  opts: NyckelExpressMiddlewareOptions
): RequestHandler {
  const { authConfig } = opts;

  return asyncHandler(async (req, res) => {
    if (!req.query.state) {
      throw new Error("Expected query to contain state!");
    }
    if (!req.query.code) {
      throw new Error("Expected query to contain code!");
    }

    const session = await req.session.get();

    try {
      const { csrfSecret } = session;
      await verifyCSRFPair({
        secret: csrfSecret,
        token: req.query.state
      });
    } catch (err) {
      await req.session.delete();
      throw err;
    }

    const data = await requestToken(
      req.query.code,
      concatUrl(req.realHost, "/auth/callback"),
      authConfig
    );
    await req.session.set(data);
    res.redirect("/");
  });
}
