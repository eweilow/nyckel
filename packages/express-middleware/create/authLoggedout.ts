import { RequestHandler } from "express";
import { NyckelExpressMiddlewareOptions } from "../middleware";
import { asyncHandler } from "../asyncHandler";

export function createLoggedOutHandler(
  opts: NyckelExpressMiddlewareOptions
): RequestHandler {
  return asyncHandler(async (req, res) => {
    await req.session.delete();
    res.redirect("/");
  });
}
