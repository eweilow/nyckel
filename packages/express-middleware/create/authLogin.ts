import { RequestHandler } from "express";
import { NyckelExpressMiddlewareOptions } from "../middleware";
import { asyncHandler } from "..";
import { authorizeUser, concatUrl } from "@nyckel/authentication";

export function createLoginHandler(
  opts: NyckelExpressMiddlewareOptions
): RequestHandler {
  const { authConfig } = opts;

  return asyncHandler(async (req, res) => {
    if (await req.user.get()) {
      // Do nothing if user is already logged in
      res.redirect("/");
      return;
    }

    const { csrfPair, redirectTo } = await authorizeUser(
      [],
      concatUrl(req.realHost, "/auth/callback"),
      authConfig
    );

    await req.session.set({
      csrfSecret: csrfPair.secret
    });

    res.redirect(redirectTo);
  });
}
