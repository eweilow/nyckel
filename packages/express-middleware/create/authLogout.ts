import { RequestHandler } from "express";
import { NyckelExpressMiddlewareOptions } from "../middleware";
import { concatUrl, logoutUser } from "@nyckel/authentication";
import { asyncHandler } from "../asyncHandler";

export function createLogoutHandler(
  opts: NyckelExpressMiddlewareOptions
): RequestHandler {
  const { authConfig } = opts;

  return asyncHandler(async (req, res) => {
    const { redirectTo } = await logoutUser(
      concatUrl(req.realHost, "/auth/loggedout"),
      authConfig
    );

    res.redirect(redirectTo);
  });
}
