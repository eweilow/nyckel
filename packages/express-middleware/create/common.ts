import { RequestHandler } from "express";
import { getSafeHostname } from "@nyckel/authentication";
import {
  enhanceRequestWithSession,
  enhanceRequestWithUser
} from "@nyckel/http-middleware";
import { NyckelExpressMiddlewareOptions } from "../middleware";

export function createCommonMiddleware(
  opts: NyckelExpressMiddlewareOptions
): RequestHandler {
  const {
    cookieName,
    secureCookies,
    sessionManager,
    trustProxyFn,
    authConfig,
    authManagementConfig
  } = opts;

  return (req, res, next) => {
    req.realHost = getSafeHostname(
      req.protocol,
      req.get("X-Forwarded-Host"),
      req.get("Host"),
      trustProxyFn
    );

    const cookie = req.cookies[cookieName];

    enhanceRequestWithSession(
      req,
      cookieName,
      cookie,
      sessionManager,
      (name, value) =>
        res.cookie(name, value, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 180,
          secure: secureCookies || req.get("X-Forwarded-Proto") === "https"
        }),
      name => res.clearCookie(name)
    );
    enhanceRequestWithUser(
      req,
      cookieName,
      cookie,
      sessionManager,
      authConfig,
      authManagementConfig
    );

    next();
  };
}
