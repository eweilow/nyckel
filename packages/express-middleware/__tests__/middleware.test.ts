import {
  nyckelExpressMiddleware,
  NyckelExpressMiddlewareOptions
} from "../middleware";
import { RequestHandler } from "express";
import express from "express";
import supertest from "supertest";
import { createCommonMiddleware } from "../create/common";
import { createLogoutHandler } from "../create/authLogout";
import { createLoggedOutHandler } from "../create/authLoggedout";
import { createCallbackHandler } from "../create/authCallback";
import { createLoginHandler } from "../create/authLogin";

function mockFactory(name: string) {
  return () => {
    return {
      [name]: jest.fn(() => {
        const handler: RequestHandler = (req, res) => {
          res.end(name);
        };
        return handler;
      })
    };
  };
}
jest.mock("../create/common", () => {
  return {
    createCommonMiddleware: jest.fn(() => {
      const handler: RequestHandler = (req, res, next) => {
        res.write("createCommonMiddleware:");
        next();
      };
      return handler;
    })
  };
});
jest.mock("../create/authLogin", mockFactory("createLoginHandler"));
jest.mock("../create/authCallback", mockFactory("createCallbackHandler"));
jest.mock("../create/authLoggedout", mockFactory("createLoggedOutHandler"));
jest.mock("../create/authLogout", mockFactory("createLogoutHandler"));

describe("nyckelExpressMiddleware", () => {
  it("should create a correct router", async () => {
    const opts = {} as NyckelExpressMiddlewareOptions;
    const app = express();
    app.use(nyckelExpressMiddleware(opts));
    const server = supertest(app);

    expect((await server.get("/auth/login")).text).toBe(
      "createCommonMiddleware:createLoginHandler"
    );
    expect((await server.get("/auth/callback")).text).toBe(
      "createCommonMiddleware:createCallbackHandler"
    );
    expect((await server.get("/auth/loggedout")).text).toBe(
      "createCommonMiddleware:createLoggedOutHandler"
    );
    expect((await server.get("/auth/logout")).text).toBe(
      "createCommonMiddleware:createLogoutHandler"
    );

    expect(createCommonMiddleware).toBeCalledWith(opts);
    expect(createLoginHandler).toBeCalledWith(opts);
    expect(createCallbackHandler).toBeCalledWith(opts);
    expect(createLoggedOutHandler).toBeCalledWith(opts);
    expect(createLogoutHandler).toBeCalledWith(opts);
  });

  it("returns the correct router stack", () => {
    const opts = {} as NyckelExpressMiddlewareOptions;
    const router = nyckelExpressMiddleware(opts);

    expect(
      router.stack.map(item => ({
        path: item.route ? item.route.path : null,
        regexp: item.regexp
      }))
    ).toMatchSnapshot();
  });
});
