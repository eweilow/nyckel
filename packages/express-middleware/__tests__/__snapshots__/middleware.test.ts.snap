// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nyckelExpressMiddleware returns the correct router stack 1`] = `
Array [
  Object {
    "path": null,
    "regexp": /\\^\\\\/\\?\\(\\?=\\\\/\\|\\$\\)/i,
  },
  Object {
    "path": "/auth/login",
    "regexp": /\\^\\\\/auth\\\\/login\\\\/\\?\\$/i,
  },
  Object {
    "path": "/auth/logout",
    "regexp": /\\^\\\\/auth\\\\/logout\\\\/\\?\\$/i,
  },
  Object {
    "path": "/auth/loggedout",
    "regexp": /\\^\\\\/auth\\\\/loggedout\\\\/\\?\\$/i,
  },
  Object {
    "path": "/auth/callback",
    "regexp": /\\^\\\\/auth\\\\/callback\\\\/\\?\\$/i,
  },
]
`;
